src/curl.c

```diff
@@ -169,35 +169,57 @@ size_t swupd_download_file(void *ptr, size_t size, size_t nmemb, void *userdata)
 	const char *outfile;
 	int fd;
 	FILE *f;
-	size_t written;
+	size_t written, remaining;
 
 	outfile = file->staging;
+	if (file->fd_valid) {
+		fd = file->fd;
+	} else {
+		fd = open(outfile, O_CREAT | O_RDWR | O_CLOEXEC | O_APPEND, 00600);
+		if (fd < 0) {
+			fprintf(stderr, "Cannot open file for write \\*outfile=\"%s\",strerror=\"%s\"*\\\n",
+				outfile, strerror(errno));
```

Comment:
When testing this patch for a large update of Clear Linux, I am running into this error condition repeatedly. The specific open(2) error is EMFILE (Too many open files).

I think there is a bug in the hysteresis logic where too many files are being added to the download queue and not being processed fast enough, thus reaching the max open fd limit when the files are open for the entirety of the download. I will debug this issue first.